src/neptune/internal/utils/git.py

```diff
@@ -90,3 +106,94 @@ def to_git_info(git_ref: Union[GitRef, GitRefDisabled]) -> Optional[GitInfo]:
         )
     except:  # noqa: E722
         return None
+
+
+class DiffTracker:
+    def __init__(self, repo: git.Repo):
+        self._repo = repo
+        self._head: git.HEAD = self._repo.head
+        self._upstream_commit_sha: Optional[str] = None
+
+    @property
+    def repo_dirty(self) -> bool:
+        return self._repo.is_dirty()
+
+    @classmethod
+    def from_git_ref(cls, git_ref: Union[GitRef, GitRefDisabled]) -> Optional["DiffTracker"]:
+        if git_ref == GitRef.DISABLED:
+            return None
+
+        initial_repo_path = git_ref.resolve_path()
+        if initial_repo_path is None:
+            return None
+        try:
+            repo = get_git_repo(repo_path=initial_repo_path)
+        except (NoSuchPathError, InvalidGitRepositoryError):
+            return None
+
+        return cls(repo)
+
+    @property
+    def upstream_commit_sha(self) -> Optional[str]:
+        return self._upstream_commit_sha
+
+    def get_head_index_diff(self) -> Optional[str]:
+        try:
+            return self._repo.git.diff(self._head.name)
+        except GitCommandError:
+            return
+
+    def get_upstream_index_diff(self) -> Optional[str]:
+        upstream_commit = self._get_relevant_upstream_commit()
+
+        if upstream_commit and upstream_commit != self._head.commit:
+
+            self._upstream_commit_sha = upstream_commit.hexsha
+
+            try:
+                return self._repo.git.diff(upstream_commit.hexsha)
+            except GitCommandError:
+                return
+
+    def _get_relevant_upstream_commit(self) -> Optional[git.Commit]:
+        try:
+            tracking_branch = self._repo.active_branch.tracking_branch()
+        except (TypeError, ValueError):
+            return
+
+        if tracking_branch:
+            return tracking_branch.commit
+
+        return self._search_for_most_recent_ancestor()
+
+    def _search_for_most_recent_ancestor(self) -> Optional[git.Commit]:
+        most_recent_ancestor: Optional[git.Commit] = None
+
+        try:
+            for branch in self._repo.branches:
```

Comment:
Probably `branches` are deprecated and IDE says that`self._repo.heads` is recommended.